### PR TITLE
Issue 110: Switch to all JSON data for TDFs/Stims

### PR DESCRIPTION
**Pull request by MegaGeese**
-Removed the long calls to json data
-Removed calls to length 1 arrays that were a side effect of xml2json conversion. 

closes: #110 and #289

### CODE REVIEW

File: mofacts/client/lib/router.js [hunk 1]
```diff
@@ -249,8 +249,8 @@ Router.route('/instructions', {
       console.log('No one logged in - allowing template to handle');
     } else {
       const unit = Session.get('currentTdfUnit');
-      const txt = _.chain(unit).prop('unitinstructions').first().trim().value();
-      const pic = _.chain(unit).prop('picture').first().trim().value();
+      const txt = unit.unitinstructions?.trim();
```
Comment (andrewtackett) [inline mofacts/client/lib/router.js:252]: I appreciate the use of a convenient function here but optional chaining is part of es2020 and we're trying to target es6 right now as many users may have older browsers.  You might do some research and see if we can bump that, I would say we can safely assume browser versions before 3 years ago are off the table for supporting.

Comment (andrewtackett) [reply mofacts/client/lib/router.js:252]: It may also be the case that the meteor compilation process polyfills this or otherwise negates it being an issue.  You might take a look at that as well. We may be able to tune the output to ensure support of older browsers.

File: mofacts/client/views/experiment/card.js [hunk 2]
```diff
@@ -250,10 +250,10 @@ function checkSimulation() {
     return;
   }
 
-  const setspec = Session.get('currentTdfFile').tdfs.tutor.setspec[0];
+  const setspec = Session.get('currentTdfFile').tdfs.tutor.setspec;
 
-  const simTimeout = _.chain(setspec).prop('simTimeout').intval(0).value();
-  const simCorrectProb = _.chain(setspec).prop('simCorrectProb').floatval(0.0).value();
+  const simTimeout = parseInt(setspec.simTimeout);
```
Comment (andrewtackett) [inline mofacts/client/views/experiment/card.js:255]: What happens if simTimeout isn't in the tdf?

File: mofacts/client/views/experiment/card.js [hunk 3]
```diff
@@ -250,10 +250,10 @@ function checkSimulation() {
...
+  const simCorrectProb = parseFloat(setspec.simCorrectProb);
```
Comment (andrewtackett) [inline mofacts/client/views/experiment/card.js:256]: What happens if simCorrectProb isn't in the tdf?

File: mofacts/client/views/experiment/card.js [hunk 4]
```diff
@@ -1768,7 +1766,7 @@ async function unitIsFinished(reason) {
 function getButtonTrial() {
   const curUnit = Session.get('currentTdfUnit');
   // Default to value given in the unit
-  let isButtonTrial = 'true' === _.chain(curUnit).prop('buttontrial').first().trim().value().toLowerCase();
+  let isButtonTrial = true === eval(curUnit.buttontrial);
```
Comment (andrewtackett) [inline mofacts/client/views/experiment/card.js:1769]: Don't do evals, ever.  It's a massive security implication.  Better to do a string comparison boolean operation. Also users may not type it correctly cased to eval correctly

File: mofacts/client/views/experiment/unitEngine.js [hunk 5]
```diff
@@ -1408,8 +1402,7 @@ function scheduleUnitEngine() {
           // 1 - legacy was f/b, now "b" forces a button trial
           // 2 - trial type (t, d, s, m, n, i, f)
           // 3 - location (added to qidx)
-          const groupEntry = group[index * templateSize + k];
```
Comment (andrewtackett) [inline mofacts/client/views/experiment/unitEngine.js:1411]: Why is this deleted?

File: mofacts/client/views/experiment/unitEngine.js [hunk 6]
```diff
@@ -1546,17 +1539,7 @@ function scheduleUnitEngine() {
       return settings;
     }
 
-    const rawAssess = _.safefirst(unit.assessmentsession);
```
Comment (andrewtackett) [inline mofacts/client/views/experiment/unitEngine.js:1549]: I'm not saying this is wrong, but what's the tradeoff being done here by refactoring this?

File: mofacts/client/views/home/profile.js [hunk 7]
```diff
@@ -72,7 +72,7 @@ Template.profile.events({
     selectTdf(
         target.data('tdfid'),
         target.data('lessonname'),
-        target.data('currentStimuliSetId'),
+        target.data('currentstimulisetid'),
```
Comment (andrewtackett) [inline mofacts/client/views/home/profile.js:75]: Were we not passing this correctly before? I assume it's either incorrect now or before with case mattering.

File: mofacts/server/methods.js [hunk 8]
```diff
@@ -695,8 +695,8 @@ async function getExperimentState(UserId, TDFId) { // by currentRootTDFId, not c
 }
 
 // UPSERT not INSERT
-async function setExperimentState(UserId, TDFId, newExperimentState) { // by currentRootTDFId, not currentTdfId
-  serverConsole('setExperimentState:', UserId, TDFId);
+async function setExperimentState(UserId, TDFId, newExperimentState, where) { // by currentRootTDFId, not currentTdfId
+  serverConsole('setExperimentState:', where, UserId, TDFId, newExperimentState);
```
Comment (andrewtackett) [inline mofacts/server/methods.js:699]: Good call

**Review by andrewtackett — CHANGES_REQUESTED** [review]
Fairly good overall.  Few issues here and there

File: mofacts/client/lib/router.js (hunk 1 above)
Comment (MegaGeese) [reply mofacts/client/lib/router.js:252]: It looks like optional chaining isn't supported in browser versions older than 2020, and its not supported at all in IE. 
Meteor will handle the error without crashing but it still throws an exception. Ill just change the optional chaining to something else.

File: mofacts/client/views/experiment/card.js (hunk 2 above)
Comment (MegaGeese) [reply mofacts/client/views/experiment/card.js:255]: I seem to have forgotten to add the `|| 0` part to the line

File: mofacts/client/views/experiment/card.js (hunk 3 above)
Comment (MegaGeese) [reply mofacts/client/views/experiment/card.js:256]: This is the same as simTimeout

File: mofacts/client/lib/router.js (hunk 1 above)
Comment (andrewtackett) [reply mofacts/client/lib/router.js:252]: So I know babel is somehow involved in the meteor compilation process which is a tool which has the ability to transpile from one es version to another.  Take a look at whether there's some kind of configuration option in meteor that will allow us to make use of that.  It sure would be nice to have this and null coalescing operators

File: mofacts/client/views/experiment/unitEngine.js (hunk 6 above)
Comment (MegaGeese) [reply mofacts/client/views/experiment/unitEngine.js:1549]: Since the tdf parameters dont come as single length arrays parsing them is no longer necessary. That was the point of this part of the code from what I saw.

File: mofacts/client/views/home/profile.js (hunk 7 above)
Comment (MegaGeese) [reply mofacts/client/views/home/profile.js:75]: It was not passing correctly before. `data-currentStimuliSetId` gets auto-lowercased when added to `target.data`.

File: mofacts/client/views/experiment/unitEngine.js (hunk 5 above)
Comment (MegaGeese) [reply mofacts/client/views/experiment/unitEngine.js:1411]: I think I misunderstood the point of this line. I didnt realize that a tdf can have multiple groups. Ill change this in a new commit

File: mofacts/client/lib/router.js (hunk 1 above)
Comment (MegaGeese) [reply mofacts/client/lib/router.js:252]: Its supported in babel 7.8 but from what I can see we are on 7.7

Comment (andrewtackett) [reply mofacts/client/lib/router.js:252]: We should be able to bump that probably.  Might have to bump our meteor version a bit

**Review by andrewtackett — APPROVED** [review]
